connector/connector_ldap.go

```diff
@@ -210,18 +231,131 @@ func (c *LDAPConnector) Register(mux *http.ServeMux, errorURL url.URL) {
 }
 
 func (c *LDAPConnector) Sync() chan struct{} {
+	go func() {
+		for {
+			alive, killed := c.idp.ldapPool.CheckConnections()
+			if alive > 0 {
+				log.Infof("Connector ID=%v idle_conns=%v", c.id, alive)
+			}
+			if killed > 0 {
+				log.Warningf("Connector ID=%v closed %v dead connections.", c.id, killed)
+			}
+			time.Sleep(c.idp.ldapPool.PoolCheckTimer)
+		}
+	}()
 	return make(chan struct{})
 }
 
 func (c *LDAPConnector) TrustedEmailProvider() bool {
 	return c.trustedEmailProvider
 }
 
+// A LDAPPool is a Connection Pool for LDAP connections
+// Initialize exported fields and use Acquire() to get a connection.
+// Use Put() to put it back into the pool.
+type LDAPPool struct {
+	m              sync.Mutex
+	conns          map[*ldap.Conn]struct{}
+	MaxIdleConn    int
+	PoolCheckTimer time.Duration
+	ServerHost     string
+	ServerPort     uint16
+	UseTLS         bool
+	UseSSL         bool
+	TLSConfig      *tls.Config
+}
+
+// Acquire removes and returns a random connection from the pool. A new connection is returned
+// if there are no connections available in the pool.
+func (p *LDAPPool) Acquire() (*ldap.Conn, error) {
+	p.m.Lock()
+	for conn := range p.conns {
+		delete(p.conns, conn)
+		p.m.Unlock()
```

Comment:
Could of course also do the actual lifting in a private function with Lock() + defer Lock in place, and do a call to that + return p.ldapConnect() in the public function.